sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc

```diff
@@ -0,0 +1,522 @@
+= sycl_ext_oneapi_group_load_store
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+:language: {basebackend@docbook:c++:cpp}
+
+== Notice
+
+[%hardbreaks]
+Copyright (c) 2024 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties].
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Overview
+
+This extension defines free functions for load/store operations within the
+group scope.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GROUP_LOAD_STORE` to one of the values defined in the
+table below. Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Load API
+
+==== Single Value Overload
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load scalar
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = empty_properties_t>
+void group_load(Group g, InputIteratorT in_iter,
+                      OutputT& out, Properties = {});
+}
+----
+
+_Preconditions_: `in_iter` must be the same for all work-items
+in the group.
+
+._Constraints_:
+* Only available if `Group` is a work-group or sub-group.
+* `InputIteratorT` suppose to be a random access iterator.
+* Value type of `InputIteratorT` must be convertible to `OutputT`.
```

Comment:
I was thinking about the implementation a bit.  I presume these functions will use **OpSubgroupBlockReadINTEL** and **OpSubgroupBlockWriteINTEL** from [SPV_INTEL_subgroups](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_subgroups.html).

When the type is not a fundamental scalar (e.g. a class), I presume you will just do a bunch of those operations to load or store the object piecemeal.  However, this assumes that the type is trivially copyable.  Do we need to add a constraint like:

> The value type of `InputIteratorT` and the type `OutputT` must both be trivially copyable.

Reply:
I think being default-constructible would also help. So that I could do

```
using value_type = std::iterator_traits<InputIteratorT>::value_type;
static_assert(sizeof(value_type) == 6); // For weird types.
value_type tmp[NElems];
uintN x = /*spirv-block-load*/;
ushortN y = /*spirv-block-load*/;

memcpy(tmp, &x, ...);
memcpy(tmp + /* ... */, &y, ...);
```

But we can update the spec wording once we have implementation PR that makes use of that.

Reply:
Or maybe I could do

```
char storage[size_in_bytes];
auto tmp_ptr = reinterpret_cast<value_type*>(storage);
```

and it won't be necessary to default-construct...

Reply:
You should not do the `reinterpret_cast` like this because this is undefined behavior by C++ rules.  In fact, this code is likely to break if the default constructor is not trivial.  If we need the value type of `InputIteratorT` to be default constructible, we should add that as a constraint.

Reply:
@aelovikov-intel and I chatted about this.  There are two ways he can write the algorithm, which have slight different requirements on the type:

1. He can assume that the type is "implicit lifetime", or
2. He can assume that the type is both "trivially copyable" and "default constructible".

The two options differ in the case when the default constructor is not trivial but some other constructor (like the copy constructor) is trivial.  We think that either implementation will be performant for types that don't do weird things in their constructors.

The algorithm in option 2 will construct a temporary object and then memcpy data over the object.  The default constructor for some types (e.g. `std::complex`) value-initializes its fields.  We assume that the optimizer will remove this initialization code since the contents of the object will be immediately overwritten with the memcpy, but Andrei will verify.

It seems to me that the constraints in option 2 are easier to explain to users, and it's also nice that there are C++17 traits that correspond to these constraints.  (The trait `std::is_implicit_lifetime` isn't added until C++23.)  Therefore, I think we should go with option 2, and thus we should add a new constraint to these functions saying that the type must be default constructible.

Did I get that right, @aelovikov-intel?

One thing we need clarification on, Andrei -- is this a constraint on `OutputT` or on the value type of `InputIteratorT`?

Reply:
> Did I get that right, @aelovikov-intel?

Yes! Also, implicit liftime approach is best implementable using C++23's `start_lifetime_as` which is another argument in favor of `default_constructible` trait.

> One thing we need clarification on, Andrei -- is this a constraint on OutputT or on the value type of InputIteratorT?

I think it's both `InputIteratorT` and `OutputIteratorT` for stores:

```
  bytes = block_load
  InputIteratorT::value_type tmp[N]; // default_constructible
  memcpy // trivially_copyable is used here and implicitly in block_load
  out[i] = tmp[i] for ...
```

For store that would be

```
  OutputIteratorT tmp[N]; // default_constructible
  tmp[i] = in[i] for ...
  block_store  // trivially_copyable is implicitly used here
```

Reply:
Great, added! Am I right what we don't need it for a scalar version (or should I still add it for consistency)?

Reply:
We have "scalar" block load in SPIR-V, so I expect that it is needed.

Reply:
added everywhere + also for both types (input/output)